### PR TITLE
Improve error message for unresolved toolchains

### PR DESCRIPTION
**Pull request by cpsauer**
Hi wonderful Bazel folks,

I happened to look back at https://github.com/bazelbuild/bazel/issues/10134#issuecomment-1012734198, and https://github.com/bazelbuild/bazel/issues/10134 more generally, and noticed that people had been reacting, requesting a change I'd floated, so I PR'd it. In brief, this PR proposes a fix that aims to resolve the issues that lead to user frustration there.

#### Please do read that issue for context before reviewing, but first, briefly some context around it:
**Background:** It's common to need to debug toolchain matching problems when getting set up with Bazel platforms. 
**Problem:** The current error message doesn't point users to Bazel's (great!) debugging mechanism (`--toolchain_resolution_debug`). Instead the error links people to what's more likely to be a red herring (that issue, https://github.com/bazelbuild/bazel/issues/10134), leading to lots of user frustration there. The original message seems like it was probably the right one back when flipping the flag seemed imminent, but the general consensus now seems to be that the link wasn't helpful and should be removed. Besides some flaming, a bunch of people had 👍🏻'd or ❤️'d the `--toolchain_resolution_debug` debugging actions I'd posted to indicate that they wanted it added to the error message. (Those reacts are what I'd happened to see when someone posted on the issue, sending a notification.) 
**Solution:** So this PR changes the error message to (1) add actionable advice to use the debugging flag and (2) remove the link to the issue. [Minor: It also makes capitalization and punctuation consistent between the error cases.]

I'd love your thoughtful consideration and review!

Thanks, wonderful Bazel folks, for all you do!

Cheers,
Chris
(ex-Googler)

P.S. If you'd rather leave the link to #10134 in the error message--and just add the actionable debugging suggestion, we could totally do that, too. I agonized a bit over whether to remove the link, before deciding to because (1) it seemed pretty unlikely that someone would flip the flag given its status, (2) people had explicitly suggested it be removed and (3) there's the opportunity to add instructions if and when the flag is picked back up. All that said, if you'd like to add the messaging back in, we could totally just append it to the error message! (I've checked the "allow edits by maintainers" box, too, so you can just do so and merge!) Might make sense to wordsmith a little at the same time, to reduce confusion? And also, if we re-add it, should the condition be instead whether the cpp toolchain type is *anywhere* in the list?

### CODE REVIEW

**Comment by cpsauer** [comment]
cc'ing @gregestren because we've interacted before on platforms stuff, and @lberki and @ahumesky since they were active on that other issue.
{please coach me if that's not right}

**Comment by gregestren** [comment]
Thanks for this contribution @cpsauer. I'm digging out of an inbox hole but this remains in my inbox and I'll get to it with a little bit of patience.

FWIW I think this is entirely orthogonal but you still might find it interesting: https://github.com/bazelbuild/bazel/issues/14726

**Comment by cpsauer** [comment]
You're very welcome. Thank _you_. 
(Have a newfound appreciation for the inbox hole...being, newly, [a Bazel tool maintainer](https://github.com/hedronvision/bazel-compile-commands-extractor) and having to move fast to keep up.)

Thanks also for the crosslink. Haven't delved too deep into implementing toolchains yet--just the basics of selecting and using. LMK if you think these will collide somehow.

-Chris

**Comment by gregestren** [comment]
This all looks good to me, and I think your reasons are sound. I definitely support more generically actionable messaging (beyond just C++). 

I really don't know if the old link is still meaningfully relevant. Given no clear argument that it is, I'm happy to exclude it by default.

The only nitty thing I wonder about is I think of `--toolchain_resolution_debug` are more of a power user feature, i.e. people who are actively working designing toolchains and rules. For regular users who run a build and get that error I suspect they'd find the output baffling. What about also an actionable reference to the basic documentation on platform/toolchains to at least give regular users an anchor point for framing followup help requests?

CC @katre just FYI.

**Comment by gregestren** [comment]
Like https://bazel.build/concepts/platforms-intro?

**Comment by cpsauer** [comment]
Thanks, @gregestren! Really appreciate your thoughtfulness and your getting back to me.

Sounds like we're well aligned on shipping this. Awesome.

On the link: Happy to add if you'd like. But I do earnestly think that users will be ready for `--toolchain_resolution_debug` by the time they get here. `--toolchain_resolution_debug` was, at least, fantastically helpful to me at this stage, and seems to have been also for the other folks [encountering the error](https://github.com/bazelbuild/bazel/issues/10134#issuecomment-1012734198). My mental model is that you'd very likely arrive at that error because you've already read about platforms/toolchains and have been fiddling around with them, triggering that error upon making a mistake or encountering out of date docs or a bug. I'd anticipate that most users would know that the action they just took caused the error, but perhaps not what platform value they should have specified instead. And `--toolchain_resolution_debug` is awesome for helping them discover that. Plus, the main docs are pretty googlable and discoverable, but that flag, I think, much less so. 

All that said, I'll of course happily defer to you if you still think we should add the link! In that case, what about appending an additional line something like: "If platforms or toolchains are a new concept to you, we'd encourage reading https://bazel.build/concepts/platforms-intro." ?

**Comment by gregestren** [comment]
I imagine (hope) that over time it's more likely that novice users will use platform-enabled builds, even if they don't realize it (i.e. just by using a rule set the implicitly use toolchains). I think a concise link would be helpful. Although it could be in `--toolchain_resolution_debug`'s output if we don't want to spam the user error.

File: src/test/shell/integration/toolchain_test.sh
Comment (cpsauer) [reply src/test/shell/integration/toolchain_test.sh:439]: Ah, cool. Thanks, John. Will slim down to just the "To debug" part, per your suggestion.

[I see the grep-based implementation now, which precludes having newlines in the pattern--and while I'd aimed for completeness before, I suppose that line does fully test the match. Sorry you saw the test failures--figured I'd just quickly run the change by CI and then tap back today after it finished.]

Comment (cpsauer) [reply src/test/shell/integration/toolchain_test.sh:439]: Err, actually, if we're just looking to check that there's a mismatch error with the right toolchain type, probably I should change the expect_log to just be the first error line. Same spirit, but more directly checking the error. Hopefully that's okay.

**Comment by cpsauer** [comment]
Great! New commit up, hopefully responding well to your good feedback.

@gregestren, sounds like you'd still like that link: It's in--with a minor wording improvement from what I'd put up before.

**Review by katre — APPROVED** [review]
LGTM, let's see if Greg has anything left to say.

**Review by gregestren — APPROVED** [review]
LGTM. Thanks, Chris.

**Comment by gregestren** [comment]
Strange: on merge I'm getting CI failures like https://storage.googleapis.com/bazel-untrusted-buildkite-artifacts/c0202940-cb70-42de-816e-a9b3fa85cf83/src%5Ctest%5Cjava%5Ccom%5Cgoogle%5Cdevtools%5Cbuild%5Clib%5Cskyframe%5CSkyframeTests%5Cshard_15_of_20%5Ctest_attempts%5Cattempt_1.log.

which correlates to https://github.com/bazelbuild/bazel/blob/e6fbe5b6cbc33375731bf0d6315fb31264211a80/src/test/java/com/google/devtools/build/lib/skyframe/ToolchainResolutionFunctionTest.java#L192.

Why didn't the PR checks catch this?

**Comment by katre** [comment]
Was the PR not merged to head?

**Comment by cpsauer** [comment]
@gregestren, looks like it's a new test @katre added several days ago, so it wasn't on the feature branch corresponding to this PR.

**Comment by cpsauer** [comment]
How about if I merge in master and fix?
Also, after working on some Windows clangd things, I think we should use double quotes in the flag in the error message to support Windows CMD shell, so I'll make that change, too.

**Comment by cpsauer** [comment]
Great. Done. Tests rerunning!

**Comment by cpsauer** [comment]
And we're green! Good to merge?

**Comment by gregestren** [comment]
Trying again. Thanks, Chris.

**Comment by brentleyjones** [comment]
@bazel-io flag

**Comment by ckolli5** [comment]
@bazel-io fork 5.2.0

**Pull request closed without merging**
